### PR TITLE
Add signature for Error.prepareStackTrace() to Node v4, 6, 7, 8

### PR DESCRIPTION
**Pull request by hmak3d**
# GrabCAD notes

https://grabcad.atlassian.net/browse/GC-42917

This internal PR is to merge the branch into master on the _fork_.
There will be an future/public PR to move the changes from the _fork_ back upstream to the official DefinitelyTyped repo.

The changes to v4, 6, 7, 8 are all identical across all versions (add to ErrorConstructor + add tests).

For rationale of changes, see https://github.com/v8/v8/wiki/Stack%20Trace%20API#customizing-stack-traces  We make use of it in our LoggingService.

Following passes on branch:
* `npm run test node`
* `npm run lint node`

# Standard stuff for all DefinitelyTyped PRs

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/node/index.d.ts [hunk 1]
```diff
@@ -311,6 +316,22 @@ declare namespace NodeJS {
...
+        getFunction: Function; // returns the current function
+        getFunctionName: Function; // returns the name of the current function, typically its name property. If a name property is not available an attempt will be made to try to infer a name from the function's context.
+        getMethodName: Function; // returns the name of the property of this or one of its prototypes that holds the current function
+        getFileName: Function; // if this function was defined in a script returns the name of the script
+        getLineNumber: Function; // if this function was defined in a script returns the current line number
+        getColumnNumber: Function; // if this function was defined in a script returns the current column number
+        getEvalOrigin: Function; // if this function was created using a call to eval returns a CallSite object representing the location where eval was called
+        isToplevel: boolean; // is this a toplevel invocation, that is, is this the global object?
```
Comment (joras) [inline types/node/index.d.ts]: from the documentation it looks like these booleans are actually methods?

File: types/node/index.d.ts [hunk 2]
```diff
@@ -311,6 +316,22 @@ declare namespace NodeJS {
         new(stdout: WritableStream, stderr?: WritableStream): Console;
     }
 
+    export interface StackFrame {
+        getThis: Function; // returns the value of this
+        getTypeName: Function; // returns the type of this as a string. This is the name of the function stored in the constructor field of this, if available, otherwise the object's [[Class]] internal property.
```
Comment (joras) [inline types/node/index.d.ts]: why not just type methods as `getTypeName(): string` ?

File: types/node/index.d.ts [hunk 3]
```diff
@@ -311,6 +316,22 @@ declare namespace NodeJS {
         new(stdout: WritableStream, stderr?: WritableStream): Console;
     }
 
+    export interface StackFrame {
```
Comment (joras) [inline types/node/index.d.ts]: Why not call it `CallSite` to match v8

Comment (hmak3d) [reply types/node/index.d.ts]: Good idea.  Changed in 658b8af

File: types/node/index.d.ts (hunk 1 above)
Comment (hmak3d) [reply types/node/index.d.ts]: Good catch.  The problem was a [pre-existing one](https://github.com/GrabCAD/eagle-print/blob/6feff58/typings_custom/node_patch.d.ts#L14) that we've missed all these months, and I was just copying stuff over.  Issue is now fixed by 75dabcb

File: types/node/index.d.ts (hunk 2 above)
Comment (hmak3d) [reply types/node/index.d.ts]: I never actually verified the tighter signature in an actual Node runtime but the documentation seems pretty straightforward, so I made the changes a30bc16

**Review by joras — APPROVED** [review]

**Comment by hmak3d** [comment]
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/22010 is the new PR to merge the fork back upstream to maintainers.  Note that it has more changes.

**Pull request merged**